src/polyswarm/download.py

```diff
@@ -20,6 +21,10 @@ def download(ctx, hash_file, hash_type, hash_value, destination):
     api = ctx.obj['api']
     output = ctx.obj['output']
     hashes = utils.parse_hashes(hash_value, hash_file=hash_file, hash_type=hash_type, log_errors=True)
+
+    results = api.download(destination, *hashes)
+    utils.validate_results(results, error_codes.GENERAL_ERROR)
```

Comment:
I believe this would download the same files twice, so needs to be changed to loop over `results`.

More generally, I'm not sure if this change was in @sbnetops changes or yours, but I don't think we should validate all the results at the same time, as this effectively prevents streaming of results both to the user and other systems in long running tasks. It also means that a task where some things succeed and some fail, it will die midway through the operation, and before having output anything to the user. I'm not sure this is desired behavior in most cases, like cases where we are doing multiple searches where some fail and some succeed. The behavior of my PR for V1 is an example of what I'm talking about: https://github.com/polyswarm/polyswarm-cli/pull/10/files, where instead it aggregates the results, outputs all of them, and then just sets the appropriate return code.

If you still want to keep this pattern, `validate_results` could be a generator that wraps `results` and does the checking lazily. This would at least stream the results until a bad one occurs, but I think not exiting immediately is the right thing to do.

Reply:
I like the idea of something like `consume_results` that tries to iterate over a results generator and raises an exception in case it does not find any results. I have made comments about terminating the interpreter midway the execution.

Reply:
Even raising an exception here is problematic. As an example of what I'm talking about:

Imagine we have three hashes, `a`, `b`, and `c` that we search for in alphabetical order. `b` does not exist and will return no results.

What a user would *expect* in this case is that, in the text output format, results are displayed for `a` and `c`, and a message about not finding `b` would be displayed (and, since this is a search, a 0 exit code would be returned, since some searches matched). This should be the current behavior of the v1 code.

What I believe would happen in the current code is that the tool would exit silently with an error code, mid-execution. This very much would not match user expectations.

With what I understand to be @sbnetops proposed approach, once `consume_results` hit `b`, an exception would be raised and the application would exit inside a higher level exception handler, never displaying results for `c`. This is also not the behavior I believe one would expect. I'm not opposed to having a generic consumer of the results, but there are multiple cases here where not finding results != an error that we must immediately terminate due to.

Reply:
If we used a `consume_results` design, could we have it take an arg called `fail_fast`. When `True`, it will consider any error a failure and short-circuit to exit before processing the rest. And when `False` it will process all, regardless of failure and return the collection of results + errors.

Reply:
But this is not a `search` endpoint, it is a download endpoint. The user requested a download of something that does not exist. The issue here is that the user is asking for multiple downloads at the same time. If this was a single download, it should simple give an error. Since there are multiple downloads, we can choose to stop at an error or suppress the error. If we suppress the error, we can log it or not. I think we could also have a flag that explicit suppresses errors, so it either could stop at the first error or print only what it finds.

Some possible usages:

```python
>>> api.download(destination, good_hash, bad_hash, raise_on_error=False)
iter([LocalArtifact(good_hash)])
>>> api.download(destination, good_hash, bad_hash, raise_on_error=True)
iter([LocalArtifact(good_hash)]) then raise NotFoundException(bad_hash)
```

Even stopping only when you get to the problematic hash is complicated because internally it is doing all downloads in parallel, so one of the threads will raise an exception before the others, for instance:

```python
>>> api.download(destination, good_hash, bad_hash, good_hash2, raise_on_error=True)
iter([LocalArtifact(good_hash)]) then raise NotFoundException(bad_hash)
```

But `good_hash2` will have executed anyways since it was ran in parallel. So the alternative would probably be 

```python
>>> api.download(destination, good_hash, bad_hash, good_hash2, raise_on_error=False)
iter([LocalArtifact(good_hash), Error(bad_hash), LocalArtifact(good_hash2)])
>>> api.download(destination, good_hash, bad_hash, good_hash2, raise_on_error=True)
iter([LocalArtifact(good_hash)]) then raise NotFoundException(bad_hash)
```

As a user, I would expect exceptions to be raised if I asked for things that don't exist unless I explicit tell it to suppress the exception.

Reply:
I thought that `polyswarm-api` was returning the aggregate of all results, but looking at the code, that's not the case. Will consider all your feedback and figure out some other solution.